rand: fix seeding from a weak entropy source

mspncp opened the pull request:
The 'rand_generate' method is not well suited for being used with
weak entropy sources in the 'get_entropy' callback, because the
caller needs to provide a preallocated buffer without knowing
how much bytes are actually needed to collect the required entropy.

Instead we use the 'rand_get_seed' and 'rand_clear_seed' method
which were exactly designed for this purpose: it's the callee who
allocates and fills the buffer, and finally cleans it up again.

The 'rand_get_seed' and 'rand_clear_seed' method are currently
optional for a provided random generator. We could fall back to
using 'rand_generate' if those methods are not implemented.
However, imo it would be better to simply make them an officially
documented requirement for seed sources.

Fixes #22332




##### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->
- [ ] documentation is added or updated
- [ ] tests are added or updated

mspncp commented:
**Disclaimer**  this is the first time I'm attempting such heavy internal provider plumbing. I tried my best to follow the existing code examples and not to break anything. Please be gentle with a newbie, but review it carefully.

mspncp commented on crypto/rand/prov_seed.c:
```diff
@@ -77,6 +60,17 @@ void ossl_rand_cleanup_entropy(ossl_unused OSSL_LIB_CTX *ctx,
     OPENSSL_secure_clear_free(buf, len);
 }
 
+void ossl_rand_cleanup_user_entropy(OSSL_LIB_CTX *ctx,
+                                    unsigned char *buf, size_t len)
+{
+    EVP_RAND_CTX *rng = ossl_rand_get0_seed_noncreating(ctx);
+
+    if (rng != NULL && evp_rand_can_seed(rng))
+        evp_rand_clear_seed(rng, buf, len);
```
It's essential that the creator of the buffer is its owner and is responsible for freeing it. That's why I added the '..._cleanup_user_...' calls.

mspncp commented on crypto/evp/evp_rand.c:
```diff
@@ -680,3 +692,60 @@ int EVP_RAND_verify_zeroization(EVP_RAND_CTX *ctx)
+}
+
+
+size_t evp_rand_get_seed(EVP_RAND_CTX *ctx,
```
Not a public API (yet). Could be made public as `EVP_RAND_get_seed()` in 3.2 and backported as non-public function.

paulidale replied:
Wasn't intended to be EVP visible but the underlying calls are part of our public provider API.

I think there are a number of risks associated with making this public.  Chief among them, misuse of the material.

mspncp replied:
So why is `EVP_rand()` public and `EVP_RAND_get_seed()` not?

paulidale replied:
`EVP_RAND()`?? It's a type not a function.

The seeding is internal to the DRBGs -- never called by anything else.  The DRBG's know about a few of their parent's calls and go direct so that they can support cross provider chaining.

At least before this PR it was internal.

mspncp replied:
Sorry, I meant to write `EVP_RAND_nonce()`.

mspncp commented on crypto/evp/evp_rand.c:
```diff
@@ -680,3 +692,60 @@ int EVP_RAND_verify_zeroization(EVP_RAND_CTX *ctx)
+        ctx->meth->clear_seed(ctx->algctx, buffer, b_len);
+}
+
+void evp_rand_clear_seed(EVP_RAND_CTX *ctx,
+                         unsigned char *buffer, size_t b_len)
```
Not a public API (yet). Could be made public as `EVP_RAND_clear_seed()` in 3.2 and backported as non-public function.

mspncp commented on include/crypto/rand.h:
```diff
@@ -116,6 +116,8 @@ size_t ossl_rand_get_user_entropy(OSSL_LIB_CTX *ctx,
                                unsigned char *buf, size_t len);
+void ossl_rand_cleanup_user_entropy(OSSL_LIB_CTX *ctx,
+                                    unsigned char *buf, size_t len);
 size_t ossl_rand_get_nonce(OSSL_LIB_CTX *ctx,
                            unsigned char **pout, size_t min_len, size_t max_len,
                            const void *salt, size_t salt_len);
```
**Side note**: Unfortunately, the function signature of the `EVP_RAND_nonce()` function does not fit well for being used in the `rand_get_nonce` callbacks either. We might need a `EVP_RAND_get_nonce` with the matching signature in the future.
```
int EVP_RAND_nonce(EVP_RAND_CTX *ctx, unsigned char *out, size_t outlen);
```

levitte commented on providers/implementations/rands/seeding/rand_unix.c:
```diff
@@ -641,7 +641,7 @@ size_t ossl_pool_acquire_entropy(RAND_POOL *pool)
         int attempts = 3;
 
-        bytes_needed = ossl_rand_pool_bytes_needed(pool, 1 /*entropy_factor*/);
+        bytes_needed = ossl_rand_pool_bytes_needed(pool, 8 /*entropy_factor*/);
```
I assume this isn't meant to be part of the PR...

mspncp replied:
Yep. See commit message of e0542c2408e4ad8d70952b5e576cd73028646875

mspncp replied:
Did the rand test succeed on VMS?

mspncp replied:
> dropme! rand: simulate a weak entropy source
>
> TODO: this should be implemented as a test somehow

@paulidale @slontis I wasn't able to find an existing test which tests seeding using a provided seed source... how could such an important feature get accepted without a unit test?  🧐 

Having a seeding test for  a strong and a weak entropy source would be a good thing to have, don't you think so?

levitte replied:
I haven't tested this on VMS yet, but will do next

rsbeckerca replied:
I am going to test this on NonStop. Does this require egd for an effective test?

rsbeckerca replied:
All LGTM on multiple tests in multiple configs.

mspncp replied:
The `dropme!` commit has been dropped now.

mspncp commented on crypto/provider_core.c:
```diff
@@ -2144,6 +2153,13 @@ static void rand_cleanup_nonce(const OSSL_CORE_HANDLE *handle,
+                               unsigned char *buf, size_t len)
+{
+    ossl_rand_cleanup_user_nonce((OSSL_LIB_CTX *)core_get_libctx(handle),
+                            buf, len);
```
Last [force-push](https://github.com/openssl/openssl/compare/cb57c9a65f48830021db14892789b0688c771ff5..64d9da953a002cdc6a6e4b151837949bfe1a4d14) fixed the indentation and made minor grammatical corrections in the commit message.

levitte commented:
I've tested this on VMS.  The `test_rand` test recipe passes without fail.

mspncp commented:
> > I am going to test this on NonStop. Does this require egd for an effective test?
> 
> `05-test_rand.t` test PASS on NonStop with EGD, but had to merge with `master` to get it to build because this PR is somewhat out of date with the `master` branch. I would like to retest after the branch has been updated.

Last [force-push](https://github.com/openssl/openssl/compare/64d9da953a002cdc6a6e4b151837949bfe1a4d14..391be7311892aea1ad6985a805a8c0af90778673) was a plain rebase to the head of master branch without further code changes.

t8m commented on crypto/provider_core.c:
```diff
@@ -2238,11 +2254,13 @@ static const OSSL_DISPATCH core_dispatch_[] = {
     { OSSL_FUNC_GET_ENTROPY, (void (*)(void))rand_get_entropy },
+    { OSSL_FUNC_GET_USER_ENTROPY, (void (*)(void))rand_get_user_entropy },
     { OSSL_FUNC_CLEANUP_ENTROPY, (void (*)(void))rand_cleanup_entropy },
+    { OSSL_FUNC_CLEANUP_USER_ENTROPY, (void (*)(void))rand_cleanup_user_entropy },
     { OSSL_FUNC_GET_NONCE, (void (*)(void))rand_get_nonce },
-    { OSSL_FUNC_CLEANUP_NONCE, (void (*)(void))rand_cleanup_nonce },
-    { OSSL_FUNC_GET_USER_ENTROPY, (void (*)(void))rand_get_user_entropy },
     { OSSL_FUNC_GET_USER_NONCE, (void (*)(void))rand_get_user_nonce },
+    { OSSL_FUNC_CLEANUP_NONCE, (void (*)(void))rand_cleanup_nonce },
+    { OSSL_FUNC_CLEANUP_USER_NONCE, (void (*)(void))rand_cleanup_user_nonce },
 #endif
```
So you're adding these new core callbacks but I do not see them being used anywhere??

mspncp replied:
Good catch! I seems like I overlooked to add them to `ossl_prov_seeding_from_dispatch`:

https://github.com/openssl/openssl/blob/143ca66cf00c88950d689a8aa0c89888052669f4/providers/common/provider_seeding.c#L48-L68

mspncp replied:
I added the missing entries and split off the cleanup callbacks into a [separate commit](https://github.com/openssl/openssl/pull/22394/commits/ad112210e847f6b5b9adbcf4df6794b30939bdd8). Hope this makes the changes easier to review.

t8m commented:
> @t8m this pull request is intended to be an alternative to pull request #22342 for fixing issue #22332. Why is it triaged as a feature and not [eligble for backporting](https://github.com/openssl/openssl/pull/22342#issuecomment-1757374116) like #22342?

@paulidale's PR seems less invasive and more safe to backport to me. But I can get convinced otherwise.

mspncp commented:
The `dropme!` commit is gone, ready for final review.

mspncp commented:
> @paulidale's PR seems less invasive and more safe to backport to me. But I can get convinced otherwise.

It all depends on whether I can convince @paulidale or not :-).

mspncp commented:
One [last amendment](https://github.com/openssl/openssl/compare/ad112210e847f6b5b9adbcf4df6794b30939bdd8..2b54f735e8b4097c2aac9616b782ae1f619508a9)

mspncp commented:
@levitte I added documentation for the new methods as a [fixup commit](https://github.com/openssl/openssl/pull/22394/commits/cfef9cfbfec50472495eb00188e0646310ae3010). Please take a look at the fixup and reconfirm.

mspncp commented:
> Per request from @t8m I split this pull request into two parts. The other half is in #22423.

Pooh, those two parts did not commute very well. I hope, I didn't mess up my rebases.

This pull request is now based onto #22423, hence the draft mode.

mspncp commented:
To be sure I resurrected the branch before my operation and did a simple rebase on master: 

```
msp@debian:~/src/openssl/master$ git checkout -b check-rebase cfef9cfbfec50472495eb00188e0646310ae3010
Switched to a new branch 'check-rebase'
```

The diffs look good
```diff
msp@debian:~/src/openssl/master$ git rebase master
Successfully rebased and updated refs/heads/check-rebase.
msp@debian:~/src/openssl/master$ git diff check-rebase rand-weak-entropy-source 
diff --git a/doc/internal/man3/ossl_rand_get_entropy.pod b/doc/internal/man3/ossl_rand_get_entropy.pod
index 24acc072ba..be39369f2b 100644
--- a/doc/internal/man3/ossl_rand_get_entropy.pod
+++ b/doc/internal/man3/ossl_rand_get_entropy.pod
@@ -88,7 +88,7 @@ of bytes in I<*pout> or 0 on error.
 
 The functions ossl_rand_get_user_entropy(), ossl_rand_get_user_nonce(),
 ossl_rand_cleanup_user_entropy(), and ossl_rand_cleanup_user_nonce()
-were added in OpenSSL 3.0.12, 3.1.4 and 3.2.0.
+were added in OpenSSL 3.1.4 and 3.2.0.
 
 The remaining functions described here were all added in OpenSSL 3.0.

 ```

t8m commented:
@mspncp Can you please rebase this and take it out of draft?

mspncp commented:
done.

t8m reviewed (approved):
It took me quite a while to understand whether this cannot break backwards compatibility somehow but in the end I am confident this should be OK. However I'd like @paulidale to also take a look.

mspncp commented:
> It took me quite a while to understand whether this cannot break backwards compatibility somehow but in the end I am confident this should be OK. 

I'm glad to hear that. I perfectly understand your struggles, because I also had a hard time to figure out a way to fix this issue without breaking existing code.

> However I'd like @paulidale to also take a look.

I'd appreciate another pair of eyes, too. It would also be great if @paulidale could test whether [his fix](https://github.com/openssl/openssl/pull/21964) for [setting the entropy seed source for FIPS provider](https://github.com/openssl/openssl/issues/21909) still works as designed.

paulidale commented on crypto/evp/evp_rand.c:
```diff
@@ -680,3 +692,60 @@ int EVP_RAND_verify_zeroization(EVP_RAND_CTX *ctx)
+    return 0;
+}
+
+
```
nit: extra blank line :grin:

paulidale commented on crypto/rand/prov_seed.c:
```diff
@@ -44,31 +45,13 @@ size_t ossl_rand_get_user_entropy(OSSL_LIB_CTX *ctx,
+    else
         return ossl_rand_get_entropy(ctx, pout, entropy, min_len, max_len);
-
-    /* Determine how many bytes to generate */
```
Should we retain this code path for the case where `rng != NULL && !evp_rand_can_seed(rng)` ?  Existing seed sources are not going to support the new callbacks and will devolve to the current behaviour of gathering entropy themselves.  Fixing that was primary intent behind the earlier change.

mspncp replied:
> Should we retain this code path for the case where `rng != NULL && !evp_rand_can_seed(rng)` ? Existing seed sources are not going to support the new callbacks and will devolve to the current behaviour of gathering entropy themselves.

I thought about that option as well and discarded it. In view of what I wrote in https://github.com/openssl/openssl/pull/22394#issuecomment-1774773581 it is not necessary to 'fix' the case where the seed source doesn't implement the get_entropy hooks. It just makes things more convoluted.

>  Fixing that was primary intent behind the earlier change.

The intention was to fix the VMS low entropy seeding problem, which this pull request achieves as well.

paulidale commented on crypto/rand/prov_seed.c:
```diff
@@ -44,31 +45,13 @@ size_t ossl_rand_get_user_entropy(OSSL_LIB_CTX *ctx,
-    size_t ret;
 
-    if (rng == NULL)
+    if (rng != NULL && evp_rand_can_seed(rng))
```
We really should do something similar in the `ossl_rand_get_user_nonce()` as well.

paulidale replied:
This is still sensible I think and it's a straightforward change.

mspncp replied:
> We really should do something similar in the `ossl_rand_get_user_nonce()` as well.

I would have done it right from the beginning (you know I love consistency) and it would have been a straightforward change, if there hadn't been an awkward fracture in the API's consistency: 


This is the signature of the upcall:


```C
size_t ossl_rand_get_user_nonce(OSSL_LIB_CTX *ctx,
                                unsigned char **pout,
                                size_t min_len, size_t max_len,
                                const void *salt, size_t salt_len)
```

and the obvious implementation would have been

```C
size_t ossl_rand_get_user_nonce(OSSL_LIB_CTX *ctx,
                                unsigned char **pout,
                                size_t min_len, size_t max_len,
                                const void *salt, size_t salt_len)
{
   EVP_RAND_CTX *rng = ossl_rand_get0_seed_noncreating(ctx);

    if (rng != NULL && evp_rand_has_nonce(rng))
        return evp_rand_get_nonce(rng, pout, min_len, max_len, salt, salt_len);
    else
        return ossl_rand_get_nonce(ctx, pout, min_len, max_len, salt, salt_len);
}
```
analogous to the user seed implementation.

However, instead of a pair `rand_get_nonce`/`rand_clear_nonce` where the callee allocates and cleans up the memory,


```C
OSSL_CORE_MAKE_FUNC(size_t,rand_get_nonce,
                    (void *vctx, unsigned char **buffer,
                     size_t min_len, size_t max_len,
                     const void *salt, size_t salt_len))
OSSL_CORE_MAKE_FUNC(void,rand_clear_nonce,
                    (void *vctx, unsigned char *buffer, size_t b_len))
```

there is a only a method

https://github.com/openssl/openssl/blob/fac61ea4618c83826b51aebf03cbc2bc3ac7b8c8/include/openssl/core_dispatch.h#L492-L494

which is expected to be called twice, the first time with a NULL buffer to determine the the length of the buffer to be allocated by the caller (note the `unsigned char *out` instead of `unsigned char **pout`). Also, the `salt` and `salt_len` is missing from the arguments.

And that fracture is not easy to fix, because the method is public as `EVP_RAND_nonce()`. :-/

mspncp replied:
see also https://github.com/openssl/openssl/pull/22394#pullrequestreview-1678916810.

paulidale replied:
`ossl_rand_get_user_nonce()` hasn't been in a release and is consequently mutable at least until later today.  Although you're not proposing a change so it might be okay even so.

More over, `ossl_rand_get_user_nonce()` is only ever called by the FIPS provider (and similar 3rd party providers) which will never call `EVP_RAND_nonce()`, so again it should be okay.

Finally, we don't call `EVP_RAND_nonce()` anywhere and implementing it (very) inefficiently might be acceptable.

mspncp replied:
I did not fully understand you answer. Maybe we can talk about this briefly at the end of our OTC meeting?

paulidale replied:
Or before??

paulidale replied:
Essentially, `EVP_RAND_nonce()` is independent of `ossl_rand_get_user_nonce()`.  The API difference doesn't matter IMO.

mspncp replied:
> Or before??

_(That was not possible due to another meeting)_

mspncp replied:
@paulidale and I (and others) had a talk and I'll come with a proposal within the next days.

paulidale commented:
I think that this is a better approach to the problem than #22342 is.
It does, however, require extra new functions from seed providers to work.

t8m commented:
So my feeling is that this requires some non-trivial work and thus it is not going to be included in 3.1.4 and potentially even 3.2.0 beta1.

What about merging #22342 instead for now as a temporay fix and merging this one later?

@paulidale @mspncp ?

mspncp commented:
I believe that @paulidale is missing an important point:  I claim that every existing seed source which gets installed via `RAND_set_seed_source_type()` **must** already have the **get_entropy** callbacks implemented, otherwise it would be fundamentally broken:

That's because the seed source eventually gets instantiated as the parent of the main drbg:

https://github.com/openssl/openssl/blob/b6eb95fa4439ea6254a5330487dabb2a499fb6c8/crypto/rand/rand_lib.c#L734-L736

Which means that in `get_entropy()` callback the `drbg->parent != NULL` case  holds for the main drbg, so `drbg->parent_get_seed()` is used instead of the upcall:

https://github.com/openssl/openssl/blob/b6eb95fa4439ea6254a5330487dabb2a499fb6c8/providers/implementations/rands/drbg.c#L200-L212

And if `drbg->parent_get_seed == NULL` it would fail. Is my reasoning correct?

As far as I understand it, things are actually quite the other way around: the get_entropy calls are the interface which was originally intended (by NIST SP 800-90A) for seeding and implemented that way in 1.1.1. Your upcall is deviating from that by 'misusing' the generate call, which causes your problems with low entropy seed sources. So imho #22342 doesn't fix the problem, just a symptom. It's better to fix it properly in 3.2 instead of waiting until people start writing independent entropy providers and need to implement workarounds.

t8m commented:
> It's better to fix it properly in 3.2 instead of waiting until people start writing independent entropy providers and need to implement workarounds.

Sure, I am talking here about 3.1.4 where we have the VMS CI regression as well and we are about to release it tomorrow!

t8m commented:
@mspncp could you please at least address the @paulidale's comments that are relevant regardless of your previous comment?

mspncp commented:
> > It's better to fix it properly in 3.2 instead of waiting until people start writing independent entropy providers and need to implement workarounds.
> 
> Sure, I am talking here about 3.1.4 where we have the VMS CI regression as well and we are about to release it tomorrow!

You did also talk about 3.2.0 beta1, though.

According to @levitte, this pull request fixes the VMS regression. And it passes all tests. So I don't see why it couldn't go into 3.1.4 as well.

mspncp commented:
> @mspncp could you please at least address the @paulidale's comments that are relevant regardless of your previous comment?

Sure, I'll take a look.

mspncp commented:
> > @mspncp could you please at least address the @pauli#dale's comments that are relevant regardless of your previous comment?
> 
> Sure, I'll take a look.
 
OTOH: we shouldn't rush this change in. Since 3.1.4 is imminent tomorrow, I would be satisfied with accepting #22342 as quick fix for 3.1.4 and getting my fix in for 3.1.5. And certainly it should be in 3.2.0 beta1. Would that be an acceptable compromise?


I'll approve #22342 to get the timer running.

t8m commented:
@mspncp I still don't see the first two comments (about the check and the extra line) being adressed. IMO those two are the only blockers to get it approved.

t8m commented:
CI is relevant @mspncp

t8m commented:
@mspncp CI still not passing.

t8m commented:
> @mspncp CI still not passing.

Hmm, I think the correct fix is to just revert the fixup changes (except for the deleted empty line) as it is apparent that clear_seed might not be implemented.

rsbeckerca commented:
Is this PR going to impact `--with-rand-seed=rdcpu` configurations?

mspncp commented:
> Is this PR going to impact `--with-rand-seed=rdcpu` configurations?

No, as far as I can tell.

paulidale commented:
> Is this PR going to impact `--with-rand-seed=rdcpu` configurations?

It shouldn't.  This is working the plumbing between the seed and its usage and not impacting the seeds themselves.

paulidale commented:
@mspncp thanks for the reeducation :)  I'd forgotten about the seeding APIs.  They are used for a DRBG to get seeding material and are not used or exposed by the EVP APIs.  FIPS convoluted things by mandating continuous tests and an external seed source.

The proper fix to the earlier issue would have been to use these as you've done here but being dense and forgetting about them led me to use the generate call instead.

I 100% prefer this to both my previous fix and to the alternative "weakness" fix.  This is done properly, the rest are kludges.

paulidale commented on providers/implementations/rands/seed_src.c:
```diff
@@ -177,33 +177,32 @@ static size_t seed_get_seed(void *vseed, unsigned char **pout,
+
+        /* xor the additional data into the output */
+        for (i = 0 ; i < adin_len ; ++i)
+            (*pout)[i % ret] ^= adin[i];
```
I'm not sure this is worthwhile.  It's not going to hurt but it might be better pre-pended/appended to the pool and processed via the DRBG seeding mechanics.

I couldn't think of any other way to use the additional input in my flawed fix.

paulidale commented on crypto/rand/prov_seed.c:
```diff
@@ -44,31 +45,13 @@ size_t ossl_rand_get_user_entropy(OSSL_LIB_CTX *ctx,
+    if (rng != NULL && evp_rand_can_seed(rng))
+        return evp_rand_get_seed(rng, pout, entropy, min_len, max_len,
+                                 0, NULL, 0);
+    else
```
Ditch this line?

paulidale replied:
We usually don't put an else after a return but it's not a big deal.

paulidale reviewed (approved):
I'd like to see the nonce change added but it's probably not critical.

mspncp commented:
My last reply is really hard to find in pr thread. See https://github.com/openssl/openssl/pull/22394#discussion_r1369686576.

levitte reviewed (approved):
Just to reconfirm, as lots of discussions have happened since last time I did this

t8m commented:
**OTC: This is approved to go in master and 3.1 immediately.**

mattcaswell commented:
Pushed to master and 3.1. There was a trivial conflict cherry-pick to 3.1 which I resolved during the merge.
